### PR TITLE
clean up local hacks

### PR DESCRIPTION
**Pull request by danwinship**

### CODE REVIEW

**Comment by danwinship** [comment]
actually
/assign dcbw
I was testing this to figure out if the loadbalancer problem was in this part of #201 or the actual rebase, and it looks like it's in the rebase part and this is fine... so this could theoretically just merge now

File: go-controller/pkg/node/gateway_localnet.go [hunk 1]
```diff
@@ -90,8 +90,9 @@ func delIptRules(ipt util.IPTablesHelper, rules []iptRule) {
 func generateGatewayNATRules(ifname string, ip net.IP) []iptRule {
 	// Allow packets to/from the gateway interface in case defaults deny
 	rules := make([]iptRule, 0)
-	// Block MCS
+	// OCP HACK: Block MCS Access. https://github.com/openshift/ovn-kubernetes/pull/170
```
Comment (trozet) [inline go-controller/pkg/node/gateway_localnet.go:93]: what is MCS? I looked in the bug at that PR it still just points the acronym.

File: go-controller/pkg/node/gateway_localnet.go [hunk 2]
```diff
@@ -271,30 +271,9 @@ func localnetIptRules(svc *kapi.Service, gatewayIP string) []iptRule {
...
-				chain: iptableNodePortChain,
-				args: []string{
-					"-d", ing.IP,
-					"-p", string(protocol), "--dport", ingPort,
-					"-j", "ACCEPT",
-				},
-			})
-		}
+		// OCP HACK: Fix Azure/GCP LoadBalancers. https://github.com/openshift/ovn-kubernetes/pull/112
+		rules = append(rules, getLoadBalancerIPTRules(svc, svcPort, gatewayIP, svcPort.NodePort)...)
```
Comment (trozet) [inline go-controller/pkg/node/gateway_localnet.go:275]: we should go back and fix this upstream, is there a github issue already upstream?

File: go-controller/pkg/cni/helper_linux.go [hunk 3]
```diff
@@ -315,16 +313,11 @@ func (pr *PodRequest) ConfigureInterface(namespace string, podName string, ifInf
...
-	err = wait.PollImmediate(100*time.Millisecond, 20*time.Second, func() (bool, error) {
-		stdout, err := ofctlExec("dump-flows", "br-int")
-		if err != nil {
-			return false, nil
-		}
-		return strings.Contains(stdout, ifInfo.IPs[0].IP.String()), nil
-	})
-	if err != nil {
+	// OCP HACK: wait for OVN to fully process the new pod
+	if err = waitForBrIntFlows(ifInfo.IPs[0].IP.String()); err != nil {
```
Comment (trozet) [inline go-controller/pkg/cni/helper_linux.go:317]: why does this exist? how is this required downstream only?

File: go-controller/pkg/node/gateway_localnet.go (hunk 1 above)
Comment (danwinship) [reply go-controller/pkg/node/gateway_localnet.go:93]: Machine Config Server. I don't know if there's a good primer on this but basically it is a security hole if pods can reach it, and they're supposed to fix it but it's complicated. (Basically, the MCS makes it easy to auto-scale-up a cluster with new nodes, but if you can pretend to be a newly-created node, you can eventually get cluster-admin credentials.)

File: go-controller/pkg/node/gateway_localnet.go (hunk 2 above)
Comment (danwinship) [reply go-controller/pkg/node/gateway_localnet.go:275]: yes, https://github.com/ovn-org/ovn-kubernetes/pull/1126. We haven't implemented it for shared gateway mode yet so it hasn't merged upstream. (Feel free to take over that PR if you understand shared gateway mode well at this point.)

**Comment by danwinship** [comment]
I'm not sure what's going on here... I think something in CI is broken. If I use clusterbot to build an image from this PR, I can launch it with no problems. But the tests are reliably failing because they never reach the point of creating worker nodes, which does not seem to be related to any ovn-kube failures

**Comment by danwinship** [comment]
hm... e2e-gcp-ovn passed that time, and e2e-aws-ovn successfully installed a cluster but flaked on a storage test. So maybe there was something infrastructural that was broken before that's fixed now?

/retest

File: go-controller/pkg/cni/helper_linux.go (hunk 3 above)
Comment (dcbw) [reply go-controller/pkg/cni/helper_linux.go:317]: @trozet I wasn't sure whether Girish would like it but it's worth proposing upstream.

Comment (trozet) [reply go-controller/pkg/cni/helper_linux.go:317]: after discussing we need to put this upstream.

**Comment by trozet** [comment]
Adding this comment for overall tracking purposes. All of the local go hacks here should go away except the MCS one, which is downstream only.

Upstream:
- https://github.com/ovn-org/ovn-kubernetes/pull/1126
- wait for br-int flows before CNI finishes

**Pull request merged**